sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/UdfProvider.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.sql;
+
+import java.lang.reflect.Method;
+import java.util.Collections;
+import java.util.Map;
+import org.apache.beam.sdk.transforms.Combine;
+
+/**
+ * Provider for user-defined functions written in Java. Implementations should be annotated with
+ * {@link com.google.auto.service.AutoService}.
+ */
+public interface UdfProvider {
+  /** Maps function names to scalar function implementations. */
+  default Map<String, Method> userDefinedScalarFunctions() {
```

Comment:
Commented on the design doc: it is really preferable not to use `Method` here, but a specialized interface that is more specific. `Method` has a bunch of things that are not relevant and also if we want to add UDF-specific methods we can't.

Reply:
I can see why you think `Method` have too many things that might not relevant. Ideally we only need users to provide function name plus function signature, which are enough to 1) offer required information to SQL parser 2) locate the right implementation in the user jar.


The reason we still use `Method`, is because that current UDF implementation asks for  a `Method`:  [1] [2]. So as we will need to construct a `Method` eventually, it doesn't seem a bad choices to expose it in the interface.   


[1]: https://github.com/apache/beam/blob/5076255fda1700f0d3ac2a9e5e73372bdf3c59dd/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/ScalarFunctionImpl.java#L96
[2]: https://github.com/apache/beam/blob/master/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/ScalarFunctionImpl.java#L111

Reply:
Using our own interface would give us more flexibility with the implementation, but if our interface was solely a means of looking up a Method, I don't see much of a difference. If we want to add UDF-specific methods, there's no guarantee we could do that in a backward-compatible way even if we used a specialized interface.

(Mostly unrelated fun fact: Intellij can autofill `this.getClass.getMethod` calls, so it's more convenient to use than I initially thought.)

Reply:
Agreed with what @ibzib has said.

My another thought is, having a Method in our interface actually is a good thing for longer term evolvement. Once the interface is adopted, it will become harder to deprecate and upgrade this interface. Thus having `Method` provides more flexibility to allow us enhance UDF implementation without changing the interface. The information (e.g. annotations) that might be not relevant now could become relevant later.

Reply:
I don't agree. If it is a simple function, a method can always be used as in `SomeClass::methodName` and this is also easier for following control flow (including for IDEs). Whenever you can use explicit method references instead of reflection, you should.

The converse is not true. If a function is provided, it cannot automatically be converted to `Method`. The class `Method` is specifically a rich representation of an aspect of a `Class` in the JVM. It is not a good class for use when you want some sort of function. It is not a good class for anything _except_ when you specifically want to manipulate the JVM's representation of a class.

If we expand the interface, we can maintain backwards-compatibility with `default` method implementation. If we need to expand it in a way where this is not possible we will need to create a new interface. If we use `Method` it is not possible to expand the interface anyhow.

I really think if you look at the pro/con of using `Method` vs a smaller interface there is not a single thing in the `pro` column.

Reply:
@kennknowles Maybe we should have an idea of what smaller interface you are thinking of? Can you give an example?Then we can compare and see pros/cons?

Reply:
Per our offline chat, we can continue discussing what a good interface should be and this PR should not be blocked. We will mark this interface as experimental and refine when necessary.

Reply:
Most important: offline I agreed that we do not need to block on the representation of a UDF as a `Method`. But I would still like to see a test case demonstrating what it is like for a user to register a UDF using a lambda / anonymous function.